contracts/liquidityMining/KyberFairLaunch.sol

```diff
@@ -0,0 +1,469 @@
+  */
+  function poolLength() external override view returns (uint256) {
+    return poolInfo.length;
+  }
+
+  /**
+  * @dev get pending reward of a user from a pool, mostly for front-end
+  * @param _pid: id of the pool
+  * @param _user: user to check for pending rewards
+  */
+  function pendingReward(uint256 _pid, address _user) external override view returns (uint256) {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][_user];
+    uint256 accRewardPerShare = pool.accRewardPerShare;
+    uint256 lpSupply = IERC20Ext(pool.stakeToken).balanceOf(address(this));
+    uint32 lastAccountedBlock = _lastAccountedRewardBlock(_pid);
+    if (lastAccountedBlock > pool.lastRewardBlock && lpSupply != 0) {
+      uint256 reward = uint256(lastAccountedBlock - pool.lastRewardBlock).mul(pool.rewardPerBlock);
+      accRewardPerShare = accRewardPerShare.add(reward.mul(1e12).div(lpSupply));
+    }
+    return user.amount.mul(accRewardPerShare).div(1e12).sub(user.rewardDebt);
+  }
+
+  /**
+  * @dev update pool reward for all pools
+  */
+  function massUpdatePools() public {
+    uint256 length = poolInfo.length;
+    for (uint256 pid = 0; pid < length; ++pid) {
+      updatePoolRewards(pid);
+    }
+  }
+
+  /**
+  * @dev update reward for one pool
+  */
+  function updatePoolRewards(uint256 _pid) public override {
+    PoolInfo storage pool = poolInfo[_pid];
+    uint32 lastAccountedBlock = _lastAccountedRewardBlock(_pid);
+    if (lastAccountedBlock <= pool.lastRewardBlock) return;
+    uint256 lpSupply = IERC20Ext(pool.stakeToken).balanceOf(address(this));
+    if (lpSupply == 0) {
+      pool.lastRewardBlock = _safeUint32(block.number);
+      return;
+    }
+    uint256 reward = uint256(lastAccountedBlock - pool.lastRewardBlock).mul(uint256(pool.rewardPerBlock));
+    pool.accRewardPerShare = _safeUint128(uint256(pool.accRewardPerShare).add(reward.mul(1e12).div(lpSupply)));
+    pool.lastRewardBlock = lastAccountedBlock;
+  }
+
+  /**
+  * @dev check if a pool of _stakeToken exists
+  * @param _stakeToken: token to check for exisiting pool
+  */
+  function isDuplicatedPool(address _stakeToken) public view returns (bool) {
+    uint256 length = poolInfo.length;
+    for (uint256 _pid = 0; _pid < length; _pid++) {
+      if (poolInfo[_pid].stakeToken == _stakeToken) return true;
+    }
+    return false;
+  }
+
+  /**
+  * @dev withdraw _amount of stakeToken from pool _pid, also harvest reward for the sender
+  */
+  function _withdraw(uint256 _pid, uint256 _amount) internal {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    require(user.amount >= _amount, 'withdraw: insufficient amount');
+
+    // update pool reward and harvest
+    updatePoolRewards(_pid);
+    _harvest(msg.sender, _pid);
+
+    user.amount = user.amount.sub(_amount);
+    user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
```

Comment:
should we replace `1e12` with `PRECISION` or creating a constant for it

Reply:
I also prefer to move the logic updating user.rewardDebt into _harvest function

Reply:
the logics are different in deposit, withdraw and harvest.
deposit: use current user's amount to harvest, then increase user's amount -> update user's reward debt
withdraw: use current user's amount to harvest, then decrease user's amount -> update user's reward debt
harvest: use current user's amount to harvest + update user's reward debt.

Hence if we put the update user's reward debt into harvest func, it needs to differentiate among 3 cases

Reply:
Ah, there is another way to do this. It is store the lastAccRewardPerShare instead of user.rewardDebt
When harvesting, just take (pool.accRewardPerShare - lastAccRewardPerShare) * user.amount / 1 e18

Reply:
each user has a different rewardDebt and it is updated for every user's action, 
if we have a global lastAccRewardPerShare, how would u calculate reward of a user given accRewardPerShare and lastAccRewardPerShare?